File: src/18-challenge/solution.ts
type Light = '.' | 'R' | 'G'

export default function fn(board: Light[][]): boolean {
	const rows = board.length
	const cols = (board[0] ?? []).length
	if (rows < 4 && cols < 4) return false

	// Horizontal
	for (const row of board) {
		const horizontalLine = row.join('')
		if (horizontalLine.match(/R{4}|G{4}/)) return true
	}

	// Vertical
	for (let col = 0; col < cols; col++) {
		let verticalLine = ''
		for (let row = 0; row < rows; row++) verticalLine += board[row][col]
		if (verticalLine.match(/R{4}|G{4}/)) return true
	}

	// Left diagonal and right diagonal
	for (let row = 0; row <= rows - 4; row++) {
		for (let col = 0; col <= cols - 4; col++) {
			let leftDiagonalLine = ''
			for (let i = 0; i < 4; i++) leftDiagonalLine += board[row + i][col + i]
			if (leftDiagonalLine.match(/R{4}|G{4}/)) return true

			let rightDiagonalLine = ''
			for (let i = 0; i < 4; i++) rightDiagonalLine += board[row + i][col + 3 - i]
			if (rightDiagonalLine.match(/R{4}|G{4}/)) return true
		}
	}

	return false
}

File: src/18-challenge/solution.py
from typing import Literal
import re

Light = Literal[".", "R", "G"]


def fn(board: list[list[Light]]) -> bool:
    rows = len(board)
    cols = len(board[0])
    if rows < 4 and cols < 4:
        return False

    # Horizontal
    for row in board:
        horizontal_line = "".join(row)
        if re.match("R{4}|G{4}", horizontal_line):
            return True

    # Vertical
    for col in range(cols):
        vertical_line = ""
        for row in range(rows):
            vertical_line += board[row][col]

        if re.match("R{4}|G{4}", vertical_line):
            return True

    # Left diagonal and right diagonal
    for row in range(rows - 3):
        for col in range(cols - 3):
            left_diagonal_line = ""
            for i in range(4):
                left_diagonal_line += board[row + i][col + i]

            if re.match("R{4}|G{4}", left_diagonal_line):
                return True

            right_diagonal_line = ""
            for i in range(4):
                right_diagonal_line += board[row + i][col + 3 - i]

            if re.match("R{4}|G{4}", right_diagonal_line):
                return True

    return False
